         /* check to see if we've gotten to a new RDN */
         if (rdn_level >= 0) {
-            if (rdn_level != entry->set) {
+            if (rdn_level != X509_NAME_ENTRY_set(entry)) {
                 /* yes, new RDN */
                 /* add old RDN to DN */
                 rdnt = PyList_AsTuple(rdn);
@@ -499,7 +499,7 @@
                     goto fail0;
             }
         }
-        rdn_level = entry->set;
+        rdn_level = X509_NAME_ENTRY_set(entry);

         /* now add this attribute to the current RDN */
         name = X509_NAME_ENTRY_get_object(entry);
@@ -602,18 +602,19 @@
             goto fail;
         }

-        p = ext->value->data;
+        ASN1_OCTET_STRING *ext_value = X509_EXTENSION_get_data(ext);
+        p = ASN1_STRING_get0_data(ext_value);
         if (method->it)
             names = (GENERAL_NAMES*)
               (ASN1_item_d2i(NULL,
                              &p,
-                             ext->value->length,
+                             ASN1_STRING_length(ext_value),
                              ASN1_ITEM_ptr(method->it)));
         else
             names = (GENERAL_NAMES*)
               (method->d2i(NULL,
                            &p,
-                           ext->value->length));
+                           ASN1_STRING_length(ext_value)));

         for(j = 0; j < sk_GENERAL_NAME_num(names); j++) {
             /* get a rendering of each name in the set of names */
@@ -1512,12 +1513,16 @@
     PySSL_BEGIN_ALLOW_THREADS
     if (proto_version == PY_SSL_VERSION_TLS1)
         ctx = SSL_CTX_new(TLSv1_method());
+    #ifndef OPENSSL_NO_SSL3
     else if (proto_version == PY_SSL_VERSION_SSL3)
         ctx = SSL_CTX_new(SSLv3_method());
-#ifndef OPENSSL_NO_SSL2
+    #endif
+    #if !defined(OPENSSL_NO_SSL2) && defined(OPENSSL_VERSION_NUMBER) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
     else if (proto_version == PY_SSL_VERSION_SSL2)
         ctx = SSL_CTX_new(SSLv2_method());
 #endif
+
+
     else if (proto_version == PY_SSL_VERSION_SSL23)
         ctx = SSL_CTX_new(SSLv23_method());
     else
@@ -1975,7 +1980,12 @@
                           PyUnicode_FSConverter, &path))
         return NULL;

+    #if OPENSSL_VERSION_NUMBER >= 0x10100000L
+    bytes = -1;
+    #else
     bytes = RAND_egd(PyBytes_AsString(path));
+    #endif
+
     Py_DECREF(path);
     if (bytes == -1) {
         PyErr_SetString(PySSLErrorObject,
--- /tmp/py326_orig/Python-3.2.6/Modules/_hashopenssl.c Sun Oct 12 15:52:03 2014
+++ /usr/restore_stripe/src/Python-3.2.6/Modules/_hashopenssl.c Wed Sep 30 21:14:05 2026
@@ -38,6 +38,14 @@

 /* EVP is the preferred interface to hashing in OpenSSL */
 #include <openssl/evp.h>
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define EVP_MD_CTX_new() EVP_MD_CTX_create()
+#define EVP_MD_CTX_free(ctx) EVP_MD_CTX_destroy(ctx)
+#else
+#define EVP_MD_CTX_new() malloc(sizeof(EVP_MD_CTX))
+#define EVP_MD_CTX_free(ctx) do { EVP_MD_CTX_cleanup(ctx); free(ctx); } while(0)
+#endif
+
 /* We use the object interface to discover what hashes OpenSSL supports. */
 #include <openssl/objects.h>

@@ -59,21 +67,20 @@
 typedef struct {
     PyObject_HEAD
     PyObject            *name;  /* name of this hash algorithm */
-    EVP_MD_CTX           ctx;   /* OpenSSL message digest context */
+    EVP_MD_CTX          *ctx;   /* OpenSSL message digest context */
 #ifdef WITH_THREAD
     PyThread_type_lock   lock;  /* OpenSSL context lock */
 #endif
 } EVPobject;

-
 static PyTypeObject EVPtype;


 #define DEFINE_CONSTS_FOR_NEW(Name)  \
     static PyObject *CONST_ ## Name ## _name_obj = NULL; \
-    static EVP_MD_CTX CONST_new_ ## Name ## _ctx; \
     static EVP_MD_CTX *CONST_new_ ## Name ## _ctx_p = NULL;

+
 DEFINE_CONSTS_FOR_NEW(md5)
 DEFINE_CONSTS_FOR_NEW(sha1)
 #ifdef _OPENSSL_SUPPORTS_SHA2
@@ -93,6 +100,11 @@
     if (retval != NULL) {
         Py_INCREF(name);
         retval->name = name;
+        retval->ctx = EVP_MD_CTX_new();
+        if (retval->ctx == NULL) {
+            Py_DECREF(retval);
+            return NULL;
+        }
 #ifdef WITH_THREAD
         retval->lock = NULL;
 #endif
@@ -101,6 +113,7 @@
     return retval;
 }

+
 static void
 EVP_hash(EVPobject *self, const void *vp, Py_ssize_t len)
 {
@@ -111,7 +124,7 @@
             process = MUNCH_SIZE;
         else
             process = Py_SAFE_DOWNCAST(len, Py_ssize_t, unsigned int);
-        EVP_DigestUpdate(&self->ctx, (const void*)cp, process);
+        EVP_DigestUpdate(self->ctx, (const void*)cp, process);
         len -= process;
         cp += process;
     }
@@ -126,7 +139,9 @@
     if (self->lock != NULL)
         PyThread_free_lock(self->lock);
 #endif
-    EVP_MD_CTX_cleanup(&self->ctx);
+    if (self->ctx) {
+        EVP_MD_CTX_free(self->ctx);
+    }
     Py_XDECREF(self->name);
     PyObject_Del(self);
 }
@@ -134,7 +149,7 @@
 static void locked_EVP_MD_CTX_copy(EVP_MD_CTX *new_ctx_p, EVPobject *self)
 {
     ENTER_HASHLIB(self);
-    EVP_MD_CTX_copy(new_ctx_p, &self->ctx);
+    EVP_MD_CTX_copy(new_ctx_p, self->ctx);
     LEAVE_HASHLIB(self);
 }

@@ -162,16 +177,16 @@
 EVP_digest(EVPobject *self, PyObject *unused)
 {
     unsigned char digest[EVP_MAX_MD_SIZE];
-    EVP_MD_CTX temp_ctx;
+    EVP_MD_CTX *temp_ctx = EVP_MD_CTX_new();
     PyObject *retval;
     unsigned int digest_size;

-    locked_EVP_MD_CTX_copy(&temp_ctx, self);
-    digest_size = EVP_MD_CTX_size(&temp_ctx);
-    EVP_DigestFinal(&temp_ctx, digest, NULL);
+    locked_EVP_MD_CTX_copy(temp_ctx, self);
+    digest_size = EVP_MD_CTX_size(temp_ctx);
+    EVP_DigestFinal(temp_ctx, digest, NULL);

     retval = PyBytes_FromStringAndSize((const char *)digest, digest_size);
-    EVP_MD_CTX_cleanup(&temp_ctx);
+    EVP_MD_CTX_free(temp_ctx);
     return retval;
 }

@@ -182,17 +197,17 @@
 EVP_hexdigest(EVPobject *self, PyObject *unused)
 {
     unsigned char digest[EVP_MAX_MD_SIZE];
-    EVP_MD_CTX temp_ctx;
+    EVP_MD_CTX *temp_ctx = EVP_MD_CTX_new();
     PyObject *retval;
     char *hex_digest;
     unsigned int i, j, digest_size;

     /* Get the raw (binary) digest value */
-    locked_EVP_MD_CTX_copy(&temp_ctx, self);
-    digest_size = EVP_MD_CTX_size(&temp_ctx);
-    EVP_DigestFinal(&temp_ctx, digest, NULL);
+    locked_EVP_MD_CTX_copy(temp_ctx, self);
+    digest_size = EVP_MD_CTX_size(temp_ctx);
+    EVP_DigestFinal(temp_ctx, digest, NULL);

-    EVP_MD_CTX_cleanup(&temp_ctx);
+    EVP_MD_CTX_free(temp_ctx);

     /* Allocate a new buffer */
     hex_digest = PyMem_Malloc(digest_size * 2 + 1);
@@ -263,7 +278,7 @@
 EVP_get_block_size(EVPobject *self, void *closure)
 {
     long block_size;
-    block_size = EVP_MD_CTX_block_size(&self->ctx);
+    block_size = EVP_MD_CTX_block_size(self->ctx);
     return PyLong_FromLong(block_size);
 }

@@ -271,7 +286,7 @@
 EVP_get_digest_size(EVPobject *self, void *closure)
 {
     long size;
-    size = EVP_MD_CTX_size(&self->ctx);
+    size = EVP_MD_CTX_size(self->ctx);
     return PyLong_FromLong(size);
 }

@@ -429,9 +444,9 @@
         return NULL;

     if (initial_ctx) {
-        EVP_MD_CTX_copy(&self->ctx, initial_ctx);
+        EVP_MD_CTX_copy(self->ctx, initial_ctx);
     } else {
-        EVP_DigestInit(&self->ctx, digest);
+        EVP_DigestInit(self->ctx, digest);
     }

     if (cp && len) {
@@ -589,11 +604,12 @@

 /* used in the init function to setup a constructor: initialize OpenSSL
    constructor constants if they haven't been initialized already.  */
+
 #define INIT_CONSTRUCTOR_CONSTANTS(NAME)  do { \
     if (CONST_ ## NAME ## _name_obj == NULL) { \
         CONST_ ## NAME ## _name_obj = PyUnicode_FromString(#NAME); \
         if (EVP_get_digestbyname(#NAME)) { \
-            CONST_new_ ## NAME ## _ctx_p = &CONST_new_ ## NAME ## _ctx; \
+            CONST_new_ ## NAME ## _ctx_p = EVP_MD_CTX_new(); \
             EVP_DigestInit(CONST_new_ ## NAME ## _ctx_p, EVP_get_digestbyname(#NAME)); \
         } \
     } \
